File: platforms/android/app/src/main/kotlin/com/voidwarp/android/core/ReceiveManager.kt
package com.voidwarp.android.core

import com.voidwarp.android.native.NativeLib
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Receiver state enum
 */
enum class ReceiverState {
    IDLE,
    LISTENING,
    AWAITING_ACCEPT,
    RECEIVING,
    COMPLETED,
    ERROR;
    
    companion object {
        fun fromInt(value: Int): ReceiverState = when (value) {
            0 -> IDLE
            1 -> LISTENING
            2 -> AWAITING_ACCEPT
            3 -> RECEIVING
            4 -> COMPLETED
            5 -> ERROR
            else -> ERROR
        }
    }
}

/**
 * Information about a pending incoming transfer
 */
data class PendingTransfer(
    val senderName: String,
    val senderAddress: String,
    val fileName: String,
    val fileSize: Long
) {
    val formattedSize: String
        get() = when {
            fileSize >= 1024 * 1024 * 1024 -> "%.2f GB".format(fileSize / 1024.0 / 1024.0 / 1024.0)
            fileSize >= 1024 * 1024 -> "%.1f MB".format(fileSize / 1024.0 / 1024.0)
            fileSize >= 1024 -> "%.1f KB".format(fileSize / 1024.0)
            else -> "$fileSize 字节"
        }
}

/**
 * Manages file receiving operations
 * 
 * Key improvements:
 * - isInitialized property to check if receiver is ready
 * - Better logging for debugging connection issues
 * - Port is immediately available after initialize() succeeds
 */
class ReceiveManager {
    
    private var receiverHandle: Long = 0
    private var pollingJob: Job? = null
    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())
    
    private val _state = MutableStateFlow(ReceiverState.IDLE)
    val state: StateFlow<ReceiverState> = _state.asStateFlow()
    
    private val _port = MutableStateFlow(0)
    val port: StateFlow<Int> = _port.asStateFlow()
    
    private val _pendingTransfer = MutableStateFlow<PendingTransfer?>(null)
    val pendingTransfer: StateFlow<PendingTransfer?> = _pendingTransfer.asStateFlow()
    
    private val _progress = MutableStateFlow(0f)
    val progress: StateFlow<Float> = _progress.asStateFlow()
    
    private val _bytesReceived = MutableStateFlow(0L)
    val bytesReceived: StateFlow<Long> = _bytesReceived.asStateFlow()
    
    private val _isInitialized = MutableStateFlow(false)
    val isInitialized: StateFlow<Boolean> = _isInitialized.asStateFlow()
    
    /**
     * Check if receiver is ready to accept connections
     */
    val isReady: Boolean
        get() = receiverHandle != 0L && _port.value > 0
    
    /**
     * Initialize the receiver - must be called before startReceiving
     * Creates the native receiver and binds to a port
     */
    fun initialize(): Boolean {
        if (receiverHandle != 0L) {
            android.util.Log.d("ReceiveManager", "Already initialized on port ${_port.value}")
            return true
        }
        
        return try {
            android.util.Log.d("ReceiveManager", "Creating native receiver...")
            receiverHandle = NativeLib.voidwarpCreateReceiver()
            
            if (receiverHandle == 0L) {
                android.util.Log.e("ReceiveManager", "Failed to create receiver (null handle)")
                _isInitialized.value = false
                return false
            }
            
            _port.value = NativeLib.voidwarpReceiverGetPort(receiverHandle)
            
            if (_port.value <= 0) {
                android.util.Log.e("ReceiveManager", "Invalid port returned: ${_port.value}")
                NativeLib.voidwarpDestroyReceiver(receiverHandle)
                receiverHandle = 0
                _isInitialized.value = false
                return false
            }
            
            _isInitialized.value = true
            android.util.Log.i("ReceiveManager", "Receiver initialized successfully on port: ${_port.value}")
            true
        } catch (t: Throwable) {
            android.util.Log.e("ReceiveManager", "Exception creating receiver: ${t.message}", t)
            _isInitialized.value = false
            false
        }
    }
    
    /**
     * Start listening for incoming transfers
     * Automatically calls initialize() if not already done
     */
    fun startReceiving() {
        try {
            // Initialize if not already done
            if (receiverHandle == 0L) {
                android.util.Log.d("ReceiveManager", "Not initialized, calling initialize()...")
                if (!initialize()) {
                    android.util.Log.e("ReceiveManager", "Failed to initialize receiver")
                    _state.value = ReceiverState.ERROR
                    return
                }
            }
            
            android.util.Log.i("ReceiveManager", "Starting receiver on port ${_port.value}...")
            NativeLib.voidwarpReceiverStart(receiverHandle)
            
            // Note: FileReceiverServer already binds to the port internally,
            // so we don't need to call voidwarpTransportStartServer here.
            // That would cause a port conflict and fail.
            _state.value = ReceiverState.LISTENING
            android.util.Log.i("ReceiveManager", "Receiver now LISTENING on port ${_port.value}")
            
            startPolling()
        } catch (t: Throwable) {
            android.util.Log.e("ReceiveManager", "Exception starting receiver: ${t.message}", t)
            _state.value = ReceiverState.ERROR
        }
    }
    
    /**
     * Stop listening
     */
    fun stopReceiving() {
        android.util.Log.d("ReceiveManager", "Stopping receiver...")
        
        pollingJob?.cancel()
        pollingJob = null
        
        if (receiverHandle != 0L) {
            try {
                NativeLib.voidwarpReceiverStop(receiverHandle)
            } catch (t: Throwable) {
                android.util.Log.w("ReceiveManager", "Error stopping receiver: ${t.message}")
            }
        }
        
        _state.value = ReceiverState.IDLE
        _pendingTransfer.value = null
        
        android.util.Log.d("ReceiveManager", "Receiver stopped")
    }
    
    /**
     * Accept the pending transfer
     */
    suspend fun acceptTransfer(savePath: String): Boolean = withContext(Dispatchers.IO) {
        if (receiverHandle == 0L) return@withContext false
        if (savePath.isBlank()) return@withContext false
        
        return@withContext try {
            _state.value = ReceiverState.RECEIVING
            
            val result = NativeLib.voidwarpReceiverAccept(receiverHandle, savePath)
            
            if (result == 0) {
                while (isActive && _state.value == ReceiverState.RECEIVING) {
                    _progress.value = NativeLib.voidwarpReceiverGetProgress(receiverHandle)
                    _bytesReceived.value = NativeLib.voidwarpReceiverGetBytesReceived(receiverHandle)
                    
                    val newState = ReceiverState.fromInt(
                        NativeLib.voidwarpReceiverGetState(receiverHandle)
                    )
                    
                    if (newState == ReceiverState.COMPLETED || newState == ReceiverState.ERROR) {
                        _state.value = newState
                        break
                    }
                    
                    delay(100)
                }
                
                _pendingTransfer.value = null
                _state.value == ReceiverState.COMPLETED
            } else {
                _state.value = ReceiverState.ERROR
                false
            }
        } catch (_: Throwable) {
            _state.value = ReceiverState.ERROR
            false
        }
    }
    
    /**
     * Reject the pending transfer
     */
    fun rejectTransfer() {
        if (receiverHandle == 0L) return
        
        try {
            NativeLib.voidwarpReceiverReject(receiverHandle)
            _pendingTransfer.value = null
            _state.value = ReceiverState.LISTENING
            startReceiving()
        } catch (_: Throwable) {
            _state.value = ReceiverState.ERROR
        }
    }
    
    private fun startPolling() {
        pollingJob?.cancel()
        pollingJob = scope.launch {
            android.util.Log.d("ReceiveManager", "Polling loop started")
            
            while (isActive) {
                if (receiverHandle == 0L) {
                    android.util.Log.w("ReceiveManager", "Polling stopped: receiver handle is null")
                    break
                }
                
                try {
                    val nativeState = NativeLib.voidwarpReceiverGetState(receiverHandle)
                    val newState = ReceiverState.fromInt(nativeState)
                    
                    if (newState != _state.value) {
                        android.util.Log.d("ReceiveManager", "State changed: ${_state.value} -> $newState")
                        _state.value = newState
                    }
                    
                    if (newState == ReceiverState.AWAITING_ACCEPT) {
                        val pending = NativeLib.voidwarpReceiverGetPending(receiverHandle)
                        if (pending != null && _pendingTransfer.value == null) {
                            android.util.Log.i("ReceiveManager", 
                                "Incoming transfer: ${pending.fileName} (${pending.fileSize} bytes) from ${pending.senderName} @ ${pending.senderAddress}")
                            
                            _pendingTransfer.value = PendingTransfer(
                                senderName = pending.senderName,
                                senderAddress = pending.senderAddress,
                                fileName = pending.fileName,
                                fileSize = pending.fileSize
                            )
                        }
                    }
                    
                    if (newState == ReceiverState.RECEIVING) {
                        _progress.value = NativeLib.voidwarpReceiverGetProgress(receiverHandle)
                        _bytesReceived.value = NativeLib.voidwarpReceiverGetBytesReceived(receiverHandle)
                    }
                    
                    // Automatically restart listening after completed transfer
                    if (newState == ReceiverState.COMPLETED) {
                        android.util.Log.i("ReceiveManager", "Transfer completed, restarting listener...")
                        _pendingTransfer.value = null
                        delay(500) // Brief pause before restarting
                        startReceiving()
                        break // Exit current polling loop, new one will start
                    }
                } catch (t: Throwable) {
                    android.util.Log.e("ReceiveManager", "Polling error: ${t.message}", t)
                    _state.value = ReceiverState.ERROR
                    break
                }
                
                delay(200)
            }
            
            android.util.Log.d("ReceiveManager", "Polling loop ended")
        }
    }
    
    /**
     * Cleanup resources
     */
    fun close() {
        stopReceiving()
        scope.cancel()
        
        if (receiverHandle != 0L) {
            try {
                NativeLib.voidwarpDestroyReceiver(receiverHandle)
            } catch (_: Throwable) {
            } finally {
                receiverHandle = 0
            }
        }
    }
}

File: platforms/windows/Core/ReceiveManager.cs
using System.Diagnostics;
using System.Runtime.InteropServices;
using VoidWarp.Windows.Native;

namespace VoidWarp.Windows.Core
{
    /// <summary>
    /// Receiver state enum matching Rust FFI
    /// </summary>
    public enum ReceiverState
    {
        Idle = 0,
        Listening = 1,
        AwaitingAccept = 2,
        Receiving = 3,
        Completed = 4,
        Error = 5
    }

    /// <summary>
    /// Pending transfer information
    /// </summary>
    public class PendingTransferInfo
    {
        public string SenderName { get; set; } = string.Empty;
        public string SenderAddress { get; set; } = string.Empty;
        public string FileName { get; set; } = string.Empty;
        public ulong FileSize { get; set; }

        public string FormattedSize => FileSize switch
        {
            >= 1024 * 1024 * 1024 => $"{FileSize / 1024.0 / 1024.0 / 1024.0:F2} GB",
            >= 1024 * 1024 => $"{FileSize / 1024.0 / 1024.0:F1} MB",
            >= 1024 => $"{FileSize / 1024.0:F1} KB",
            _ => $"{FileSize} bytes"
        };
    }

    /// <summary>
    /// Event args for state changed event
    /// </summary>
    public class ReceiverStateChangedEventArgs : EventArgs
    {
        public ReceiverState OldState { get; }
        public ReceiverState NewState { get; }

        public ReceiverStateChangedEventArgs(ReceiverState oldState, ReceiverState newState)
        {
            OldState = oldState;
            NewState = newState;
        }
    }

    /// <summary>
    /// Manages file receiving operations using the native core
    /// Implements event-driven architecture similar to Android's StateFlow pattern
    /// </summary>
    public class ReceiveManager : IDisposable
    {
        private IntPtr _receiverHandle = IntPtr.Zero;
        private CancellationTokenSource? _pollCts;
        private bool _disposed = false;
        private ReceiverState _lastState = ReceiverState.Idle;
        private PendingTransferInfo? _lastPendingTransfer;
        private bool _pendingTransferHandled = false;

        /// <summary>
        /// The port the receiver is listening on
        /// </summary>
        public ushort Port { get; private set; }

        /// <summary>
        /// Whether the receiver is initialized and ready
        /// </summary>
        public bool IsInitialized => _receiverHandle != IntPtr.Zero && Port > 0;

        /// <summary>
        /// Whether actively receiving or listening
        /// </summary>
        public bool IsReceiving => State == ReceiverState.Listening || State == ReceiverState.Receiving;

        /// <summary>
        /// Current receiver state
        /// </summary>
        public ReceiverState State
        {
            get
            {
                if (_receiverHandle == IntPtr.Zero) return ReceiverState.Idle;
                return (ReceiverState)NativeBindings.voidwarp_receiver_get_state(_receiverHandle);
            }
        }

        /// <summary>
        /// Event fired when an incoming transfer is detected
        /// </summary>
        public event Action<PendingTransferInfo>? TransferRequested;

        /// <summary>
        /// Event fired when transfer progress changes
        /// </summary>
        public event Action<float>? ProgressChanged;

        /// <summary>
        /// Event fired when a transfer completes (success or failure)
        /// </summary>
        public event Action<bool, string?>? TransferCompleted;

        /// <summary>
        /// Event fired when receiver state changes
        /// </summary>
        public event EventHandler<ReceiverStateChangedEventArgs>? StateChanged;

        /// <summary>
        /// Event fired when initialization is complete
        /// </summary>
        public event EventHandler<ushort>? Initialized;

        public ReceiveManager()
        {
            Debug.WriteLine("[ReceiveManager] Creating receiver...");
            
            _receiverHandle = NativeBindings.voidwarp_create_receiver();
            if (_receiverHandle == IntPtr.Zero)
            {
                Debug.WriteLine("[ReceiveManager] ERROR: Failed to create native receiver");
                throw new InvalidOperationException("Failed to create file receiver");
            }

            Port = NativeBindings.voidwarp_receiver_get_port(_receiverHandle);
            
            if (Port <= 0)
            {
                Debug.WriteLine("[ReceiveManager] ERROR: Invalid port returned");
                NativeBindings.voidwarp_destroy_receiver(_receiverHandle);
                _receiverHandle = IntPtr.Zero;
                throw new InvalidOperationException("Failed to bind receiver to a valid port");
            }
            
            Debug.WriteLine($"[ReceiveManager] Created successfully on port {Port}");
            
            // Fire initialized event
            Initialized?.Invoke(this, Port);
        }

        /// <summary>
        /// Start listening for incoming transfers
        /// </summary>
        public void StartReceiving()
        {
            if (_receiverHandle == IntPtr.Zero)
            {
                Debug.WriteLine("[ReceiveManager] Cannot start: handle is null");
                return;
            }

            Debug.WriteLine($"[ReceiveManager] Starting receiver on port {Port}...");
            
            NativeBindings.voidwarp_receiver_start(_receiverHandle);
            
            // Reset pending transfer state
            _lastPendingTransfer = null;
            _pendingTransferHandled = false;
            
            StartPolling();
            
            Debug.WriteLine("[ReceiveManager] Receiver started, polling active");
        }

        /// <summary>
        /// Stop listening
        /// </summary>
        public void StopReceiving()
        {
            Debug.WriteLine("[ReceiveManager] Stopping receiver...");
            
            _pollCts?.Cancel();
            _pollCts = null;

            if (_receiverHandle != IntPtr.Zero)
            {
                NativeBindings.voidwarp_receiver_stop(_receiverHandle);
            }
            
            UpdateState(ReceiverState.Idle);
            
            Debug.WriteLine("[ReceiveManager] Receiver stopped");
        }

        /// <summary>
        /// Accept the pending transfer
        /// </summary>
        public async Task<bool> AcceptTransferAsync(string savePath)
        {
            if (_receiverHandle == IntPtr.Zero)
            {
                Debug.WriteLine("[ReceiveManager] Cannot accept: handle is null");
                return false;
            }

            Debug.WriteLine($"[ReceiveManager] Accepting transfer, saving to: {savePath}");
            _pendingTransferHandled = true;
            
            return await Task.Run(() =>
            {
                int result = NativeBindings.voidwarp_receiver_accept(_receiverHandle, savePath);
                if (result == 0)
                {
                    Debug.WriteLine("[ReceiveManager] Transfer accepted and completed successfully");
                    TransferCompleted?.Invoke(true, null);
                    return true;
                }
                else
                {
                    Debug.WriteLine($"[ReceiveManager] Transfer failed with result: {result}");
                    TransferCompleted?.Invoke(false, "Transfer failed");
                    return false;
                }
            });
        }

        /// <summary>
        /// Reject the pending transfer
        /// </summary>
        public void RejectTransfer()
        {
            if (_receiverHandle == IntPtr.Zero)
            {
                Debug.WriteLine("[ReceiveManager] Cannot reject: handle is null");
                return;
            }

            Debug.WriteLine("[ReceiveManager] Rejecting transfer...");
            _pendingTransferHandled = true;
            
            var result = NativeBindings.voidwarp_receiver_reject(_receiverHandle);
            if (result != 0)
            {
                Debug.WriteLine($"[ReceiveManager] Reject failed with result: {result}");
                TransferCompleted?.Invoke(false, "Reject failed");
            }
            else
            {
                Debug.WriteLine("[ReceiveManager] Transfer rejected successfully");
            }
        }

        /// <summary>
        /// Get current pending transfer info
        /// </summary>
        public PendingTransferInfo? GetPendingTransfer()
        {
            if (_receiverHandle == IntPtr.Zero) return null;

            var pending = NativeBindings.voidwarp_receiver_get_pending(_receiverHandle);
            if (!pending.IsValid)
            {
                return null;
            }

            var info = new PendingTransferInfo
            {
                SenderName = Marshal.PtrToStringUTF8(pending.SenderName) ?? "Unknown",
                SenderAddress = Marshal.PtrToStringUTF8(pending.SenderAddr) ?? "",
                FileName = Marshal.PtrToStringUTF8(pending.FileName) ?? "unknown",
                FileSize = pending.FileSize
            };

            NativeBindings.voidwarp_free_pending_transfer(pending);
            return info;
        }

        /// <summary>
        /// Get current progress percentage
        /// </summary>
        public float GetProgress()
        {
            if (_receiverHandle == IntPtr.Zero) return 0;
            return NativeBindings.voidwarp_receiver_get_progress(_receiverHandle);
        }

        /// <summary>
        /// Get bytes received so far
        /// </summary>
        public ulong GetBytesReceived()
        {
            if (_receiverHandle == IntPtr.Zero) return 0;
            return NativeBindings.voidwarp_receiver_get_bytes_received(_receiverHandle);
        }

        private void UpdateState(ReceiverState newState)
        {
            if (_lastState != newState)
            {
                var oldState = _lastState;
                _lastState = newState;
                
                Debug.WriteLine($"[ReceiveManager] State changed: {oldState} -> {newState}");
                StateChanged?.Invoke(this, new ReceiverStateChangedEventArgs(oldState, newState));
            }
        }

        private void StartPolling()
        {
            // Cancel any existing polling
            _pollCts?.Cancel();
            _pollCts = new CancellationTokenSource();
            var token = _pollCts.Token;

            Task.Run(async () =>
            {
                Debug.WriteLine("[ReceiveManager] Polling loop started");
                
                while (!token.IsCancellationRequested)
                {
                    try
                    {
                        var state = State;
                        UpdateState(state);

                        if (state == ReceiverState.AwaitingAccept)
                        {
                            // Only fire event once per pending transfer
                            if (!_pendingTransferHandled)
                            {
                                var pending = GetPendingTransfer();
                                if (pending != null && 
                                    (_lastPendingTransfer == null || 
                                     _lastPendingTransfer.FileName != pending.FileName))
                                {
                                    _lastPendingTransfer = pending;
                                    Debug.WriteLine($"[ReceiveManager] Incoming transfer: {pending.FileName} ({pending.FormattedSize}) from {pending.SenderName}");
                                    TransferRequested?.Invoke(pending);
                                }
                            }
                        }
                        else if (state == ReceiverState.Receiving)
                        {
                            var progress = GetProgress();
                            ProgressChanged?.Invoke(progress);
                        }
                        else if (state == ReceiverState.Completed)
                        {
                            Debug.WriteLine("[ReceiveManager] Transfer completed, restarting listener...");
                            // Reset state for next transfer
                            _lastPendingTransfer = null;
                            _pendingTransferHandled = false;
                            // Restart listening
                            StartReceiving();
                            break; // Exit this polling loop, new one will be started
                        }
                        else if (state == ReceiverState.Idle || state == ReceiverState.Error)
                        {
                            // Reset pending transfer state
                            if (state == ReceiverState.Error)
                            {
                                Debug.WriteLine("[ReceiveManager] Error state detected");
                            }
                            _lastPendingTransfer = null;
                            _pendingTransferHandled = false;
                        }
                    }
                    catch (Exception ex)
                    {
                        Debug.WriteLine($"[ReceiveManager] Polling error: {ex.Message}");
                    }

                    await Task.Delay(200, token);
                }
                
                Debug.WriteLine("[ReceiveManager] Polling loop ended");
            }, token);
        }

        public void Dispose()
        {
            if (!_disposed)
            {
                Debug.WriteLine("[ReceiveManager] Disposing...");
                
                StopReceiving();
                
                if (_receiverHandle != IntPtr.Zero)
                {
                    NativeBindings.voidwarp_destroy_receiver(_receiverHandle);
                    _receiverHandle = IntPtr.Zero;
                }
                _disposed = true;
                GC.SuppressFinalize(this);
                
                Debug.WriteLine("[ReceiveManager] Disposed");
            }
        }
    }
}
